frontend/readhistory.lua

```diff
@@ -50,59 +48,48 @@ local function buildEntry(input_time, input_file)
         end,
         callback = function()
             selectCallback(input_file)
-        end
+        end,
     }
 end
 
-local function fileFirstOrdering(l, r)
-    if l.file == r.file then
-        return l.time > r.time
-    else
-        return l.file < r.file
+function ReadHistory:getIndexByFile(item_file)
+    for i, v in ipairs(self.hist) do
+        if item_file == v.file then
+            return i
+        end
     end
 end
 
-local function timeFirstOrdering(l, r)
-    if l.time == r.time then
-        return l.file < r.file
-    else
-        return l.time > r.time
+function ReadHistory:getIndexByTime(item_time)
+    local hist_num = #self.hist
+    if hist_num == 0 then
+        return 1
     end
-end
-
-function ReadHistory:_indexing(start)
-    --- @todo (Hzj_jie): Use binary search to find an item when deleting it.
-    for i = start, #self.hist, 1 do
-        self.hist[i].index = i
+    if item_time >= self.hist[1].time then
+        return 1
+    elseif item_time < self.hist[hist_num].time then
+        return hist_num + 1
     end
-end
-
-function ReadHistory:_sort()
-    local autoremove_deleted_items_from_history =
-        not G_reader_settings:nilOrFalse("autoremove_deleted_items_from_history")
-    if autoremove_deleted_items_from_history then
-        self:clearMissing()
-    end
-    table.sort(self.hist, fileFirstOrdering)
-    --- @todo (zijiehe): Use binary insert instead of a loop to deduplicate.
-    for i = #self.hist, 2, -1 do
-        if self.hist[i].file == self.hist[i - 1].file then
-            table.remove(self.hist, i)
+    -- binary search
+    local s, e, m, d = 1, hist_num
+    while s <= e do
+        m = math.floor((s + e) / 2)
+        if item_time < self.hist[m].time then
+            s, d = m + 1, 1
+        elseif item_time >= self.hist[m].time then
+            e, d = m - 1, 0
         end
     end
-    table.sort(self.hist, timeFirstOrdering)
-    self:_indexing(1)
+    return m + d
 end
 
--- Reduces total count in hist list to a reasonable number by removing last
--- several items.
 function ReadHistory:_reduce()
```

Comment:
(
I know you're not friend with comments :) and this one might not be needed as the function is small and obvious.
But it was there hurting nobody :)
(and also, given the name of the function, _reduce(), it might help not trying to apprehend it as if it meant "reduce" in the contect of functionnal programming - that I don't know much about :) - and so avoid taking the wrong path trying to figure what it does.)
)